Configure MongoDB and add a sample user model

algomaster99 opened the pull request:
### Testing database

1. Install MongoDB using the apt package manager(Debian, Ubuntu) or brew (Mac)
2. Configure your database settings in `.env` file.
  i. Create user in mongo shell.
  ii. Create a database.
3. Run `npm install`.
4. Start the server and hit `/api/users` to find the list of all users.
5. Once the server is started, a collection called `users` is already created. Add users in it using the mongo shell.

### Reviewing

- Verify the directory structure of the project till now and please give suggestions to improve it.
- Suggest any code linting improvements.
- Test the way the model has been written if it completely follows the TypeScript conventions. It is subjected to change in terms of schema.

domoritz commented:
Thanks for the PR. Before we merge this, can you help me understand why we want to use Mongo instead of let's say Elastic? We have a free instance and Blockbuilder also uses it (https://github.com/enjalot/blockbuilder). If we use Mongo, what's the best way for us to host it? 

Also, I don't yet see integration with GISTs working so maybe we should do that first?

algomaster99 commented:
@domoritz I am using MongoDB since:
1. Its ODM(Mongoose) is available and updated unlike Elasticsearch's.
2. It has good support for foreign relation. We might need it for linking user and their visualizations.

I haven't added gist. I just added `UserSchema` to get familiar with creating models. Next step would be to implement Oauth.

algomaster99 commented:
@domoritz And they use `MongoDB` for authentication. For searching only, they use ElasticSearch. We can achieve searching with Mongoose as well.

domoritz commented:
I see. Search is really the only important thing since we can store the specs themselves on GISTs. You can setup a free tier MongoDB and we can use for testing and we can upgrade to a paid one when we need it. How does that sound?

domoritz commented:
Just wondering, why don't we use Postgres? We can easily host it on our servers as well. 

The main requirement we have to search is that it works across different fields. It would be nice if we could search by different features of the visualization as well (e.g. what dataset is being used or what marktype we have)

algomaster99 commented:
@domoritz I don't mind using PostgreSQL but that would require serializing the data stored in database. Since MongoDB deals with JSON objects only, we can skip this process.

siddhant1 commented:
We can use mongo db atlas or mlab (both are database as a service) to host a database

domoritz commented:
> I don't mind using PostgreSQL but that would require serializing the data stored in database. Since MongoDB deals with JSON objects only, we can skip this process.

I don't understand why this is a concern. The nice thing about postgres is that we get a full SQL database. However, if all we store is a search index, it may not be necessary.

domoritz commented on src/controllers/user.ts:
```diff
@@ -0,0 +1,28 @@
+import BaseController from './base'
+import UserModel from '../models/user'
+
+class UserController implements BaseController {
```
Why do we need to store the users?

algomaster99 replied:
- Mapping of visualizations to users will become easier so users will easily be able to see their or someone else's visualization.
- Giving a user model will help in extending the platform to not only use GitHub gist but some other platforms like JSFiddle, etc.

domoritz commented on src/controllers/user.ts:
```diff
@@ -0,0 +1,28 @@
+
+class UserController implements BaseController {
+
+  public path = '/api/users'
```
Would it make sense to instead use graphql or is it overkill if we only do simple search?

algomaster99 replied:
@domoritz I haven't used it before but I think querying with mongoose is anyway quite easy and has a good performance so it will be fine. GraphQL just provides a more intuitive way of querying data.

domoritz replied:
Yeah, GraphQL is only an API for the client. I don't know whether it's overkill for us but it might be interesting.

siddhant1 commented:
Will we store the entire visualization file in the backend or store the text of the visualization?

algomaster99 commented:
@siddhant1 We will store the GistID in back-end. The text will be stored on GitHub itself.

domoritz commented:
> @siddhant1 We will store the GistID in back-end. The text will be stored on GitHub itself.

Now quite since we want to support search and therefore we need to have some parts of the visualization in the database as well.
